torch/csrc/jit/fuser/executor.cpp

```diff
@@ -305,23 +305,30 @@ bool runFusion(
     return i.toTensor();
   });
 
-  // Determines device to dispatch to
-  // Note assumes all inputs are on at most one GPU device and/or the CPU
-  int32_t device = kCPUDevice;
-  for (const auto& t : inputs) {
-    const auto cur_device = t.device().index();
-    if (cur_device < 0) continue;
-    if (device < 0) device = cur_device;
-    else JIT_ASSERT(device == cur_device);
+  // Determines device to dispatch to. If there's a device mismatch in the inputs,
+  // we use the fallback (which should give a nice error message).
```

Comment:
What if part of the FusionGroup could have been fused? (Consider a fusion group with some CPU inputs and some CUDA:0 inputs). Are there plans to slice these things apart?

Reply:
Well if you look at our current fusion groups, there are no operations that could cause cross-device transfers. So the subgraphs that we would fuse would error out if you substituted some inputs to be on different devices. Instead of checking those things statically, we do a dynamic check now, and use the fallback to give a nicer error message if we detect that inputs are on different devices.

Reply:
Good point on the cross-device transfers